import "rec/command/Command.proto";
import "rec/command/map/CommandMap.proto";

package rec.slow;

message SlowCommand {
  enum Type {
    NONE = 0;

    ABOUT_THIS_PROGRAM = 58;
    ADD_LOOP_POINT = 1;
    AUDIO_PREFERENCES = 2;

    CHECK_FOR_UPDATES = 68;
    CLEAR_KEYBOARD_MAPPINGS = 91;
    CLEAR_MIDI_MAPPINGS = 92;
    CLEAR_LOOPS = 4;
    CLEAR_NAVIGATOR = 3;
    CLEAR_ALL_SETTINGS = 79;
    CLEAR_SETTINGS_FOR_THIS_TRACK = 80;
    CLEAR_SAVED_FILE_DATA = 5;
    CLOSE_FILE = 6;
    COPY_ALL_LOOP_POINTS = 95;

    DIM_VOLUME_TOGGLE = 7;

    EJECT_CDS = 8;
    EXPORT_KEYBOARD_MAPPINGS = 82;
    EXPORT_MIDI_MAPPINGS = 83;
    EXPORT_SETTINGS = 75;

    IMPORT_KEYBOARD_MAPPINGS = 84;
    IMPORT_MIDI_MAPPINGS = 85;
    IMPORT_SETTINGS = 76;
    INVERT_LOOP_SELECTION = 9;

    JUMP = 100000;
    JUMP_SELECTED  = 110000;

    KEYBOARD_MAPPINGS = 12;

    LOOP_NEXT_SEGMENT = 94;

    MIDI_MAPPINGS = 13;
    MODE_ADD_LOOP_POINT = 63;
    MODE_DRAG = 60;
    MODE_SET_TIME = 61;
    MODE_ZOOM_IN = 62;
    MUTE_VOLUME_TOGGLE = 14;

    NUDGE_BACKWARD = 89;
    NUDGE_FORWARD = 90;
    NUDGE_BEGIN_LEFT = 15;
    NUDGE_BEGIN_RIGHT = 16;
    NUDGE_END_LEFT = 17;
    NUDGE_END_RIGHT = 18;
    NUDGE_SPEED_DOWN = 86;
    NUDGE_SPEED_UP = 87;
    NUDGE_VOLUME_DOWN = 19;
    NUDGE_VOLUME_UP = 20;

    OPEN = 21;
    OPEN_MANUAL = 65;
    OPEN_PREVIOUS_FILE = 78;
    OPEN_SLOWGOLD_DIRECTORY = 74;

    PASTE_OVER_LOOP_POINTS = 96;

    RECENT_FILES = 220000;
    REDO = 23;
    REQUEST_SUPPORT = 66;

    RESET_GAIN_TO_UNITY = 24;

    SAVE_FILE = 69;
    SAVE_FILE_SELECTION = 70;

    SELECT = 250000;
    SELECT_ONLY = 260000;
    SET_LANGUAGE = 930000;
    SET_SAVE_FORMAT = 710000;

    // TODO: delete this.
    SET_SAVE_AS_AIFF = 97;
    SET_SAVE_AS_FLAC = 98;
    SET_SAVE_AS_OGG = 99;
    SET_SAVE_AS_WAV = 100;

    TOGGLE_ADVANCED_MENUS = 59;
    TOGGLE_AUTOMATIC_UPDATES = 67;
    TOGGLE_FOLLOW_CURSOR = 43;
    TOGGLE_GRID_DISPLAY = 27;
    TOGGLE_COMMAND_BAR_AT_LEFT = 54;
    TOGGLE_COMMAND_BAR_AT_TOP = 55;
    TOGGLE_FILE_OPEN_TREE_DISPLAY = 72;
    TOGGLE_MODES_AT_LEFT = 56;
    TOGGLE_MODES_AT_TOP = 57;
    TOGGLE_PARALLEL_WAVEFORMS = 42;
    TOGGLE_PREFADER_LEVELS = 46;
    TOGGLE_SELECTION = 410000;
    TOGGLE_SHOW_ABOUT_WINDOW_AT_STARTUP = 81;
    TOGGLE_SHOW_CURSOR_LABELS = 51;
    TOGGLE_SHOW_HELP_PANE = 49;
    TOGGLE_SHOW_LABELS_AT_TOP = 52;
    TOGGLE_SHOW_MASTER_TUNE = 77;
    TOGGLE_SHOW_SELECTION_BUTTONS = 50;
    TOGGLE_SHOW_TIMES_AT_TOP = 53;
    TOGGLE_SHOW_TOOLTIPS = 48;
    TOGGLE_START_STOP = 28;
    TOGGLE_STRETCH_ENABLE = 29;
    TOGGLE_TRANSFORM_ENABLE = 88;
    TOGGLE_WHOLE_SONG_LOOP = 30;

    TREE_CLOSE = 31;
    TREE_DOWN = 32;
    TREE_LEFT = 33;
    TREE_OPEN = 34;
    TREE_RIGHT = 35;
    TREE_UP = 36;

    UNDO = 37;
    UNSELECT = 380000;

    WHATS_NEW_PAGE = 73;

    ZOOM_IN = 39;
    ZOOM_OUT = 40;
    ZOOM_OUT_FULL = 64;
    ZOOM_TO_SELECTION = 45;

    FIRST_FREE_TYPE = 101;

    // Pre-defined by Juce.
    QUIT = 0x1001;
    DEL = 0x1002;
    CUT = 0x1003;
    COPY = 0x1004;
    PASTE = 0x1005;
    SELECT_ALL = 0x1006;
    DESELECT_ALL = 0x1007;
  };

  optional Type type = 1;
  optional rec.command.Command command = 2;
};

message SlowCommands {
  repeated SlowCommand slow_command = 1;
};

message SlowCommandMapEntry {
  optional SlowCommand.Type command = 1;
  optional rec.command.CommandMapEntry entry = 2;
};

message SlowCommandMapProto {
  repeated SlowCommandMapEntry slow_entry = 1;
};

message SlowMenuEntry {
  repeated SlowCommand.Type command = 1;
  optional string submenu = 2;
  optional string callout_function = 3;
};

message SlowMenu {
  optional command.Description description = 1;
  optional string extends = 2;
  repeated SlowMenuEntry entry = 3;
};

message SlowMenus {
  repeated SlowMenu menu = 1;
};
